contracts/modules/TransferManager/GeneralTransferManager.sol

```diff
@@ -168,16 +145,19 @@ contract GeneralTransferManager is ITransferManager {
                 //Anyone on the whitelist can transfer, regardless of time
                 return (_onWhitelist(_to) && _onWhitelist(_from)) ? Result.VALID : Result.NA;
             }
+            if (_from == issuanceAddress && (whitelist[_to].canBuyFromSTO == 0) && _isSTOAttached()) {
+                return Result.NA;
+            }
             if (allowAllWhitelistIssuances && _from == issuanceAddress) {
-                if (!whitelist[_to].canBuyFromSTO && _isSTOAttached()) {
-                    return Result.NA;
-                }
                 return _onWhitelist(_to) ? Result.VALID : Result.NA;
             }
+            if (_from == issuanceAddress) {
+                return (_onWhitelist(_to) && _adjustTimes(whitelist[_to].toTime) <= uint64(now)) ? Result.VALID : Result.NA;
+            }
             //Anyone on the whitelist can transfer provided the blocknumber is large enough
             /*solium-disable-next-line security/no-block-members*/
-            return ((_onWhitelist(_from) && whitelist[_from].fromTime <= now) &&
-                (_onWhitelist(_to) && whitelist[_to].toTime <= now)) ? Result.VALID : Result.NA; /*solium-disable-line security/no-block-members*/
+            return ((_onWhitelist(_from) && _adjustTimes(whitelist[_from].fromTime) <= uint64(now)) &&
+                (_onWhitelist(_to) && _adjustTimes(whitelist[_to].toTime) <= uint64(now))) ? Result.VALID : Result.NA; /*solium-disable-line security/no-block-members*/
```

Comment:
Shouldn't we be checking `whitelist[_to].toTime` AND `whitelist[_from].fromTime` for both `_from` and `_to`?
The current implementation will allow the transfer to happen when whitelist period of `_to` hasn't even started and whitelist period of `_from` has already expired.
Also, maybe we should remove forcing the sender to be on the whitelist.

Reply:
> Shouldn't we be checking whitelist[_to].toTime AND whitelist[_from].fromTime for both _from and _to?

@maxsam4 we are checking it but it can be overwrite if any offset time is set by the issuer. 
 

> The current implementation will allow the transfer to happen when whitelist period of _to hasn't even started and whitelist period of _from has already expired.

As per the current implementation it is fine because whiteliste period is started when the user get whitelisted and remains whitelisted until its `expiryTime` is < `now`, that we are checking in the `_onWhiteList()` function. So both users will be in the white list and the second part is to check the buy lockup period of  `_to` and sell lockup period of `_from`. that we are covering in the `_adjustTimes(whitelist[_to].toTime) <= uint64(now))) ` &  _adjustTimes(whitelist[_from].fromTime) <= uint64(now)). 

> Also, maybe we should remove forcing the sender to be on the whitelist.

That can be done by assuming that  `_to` address is always be whitelisted.

Reply:
Ah, thanks for the explanation. Correct me if I'm wrong:
1) kyc expiry time is different from transfer restriction time. 
2) `toTime` and `fromTime` are not starting and ending time of allowance/restriction but times up till which an investor and `receive` or `send` tokens. 

We should consider adding a comment about it in the code

Reply:
Yes, you are right for both points,

1. KYC expiry time is more of validating time for your identity after expiration may be an investor turn from accredited to the non-accredited investor or something else because every specific provide by him/her are changeable by time.
2. `toTime`  and `fromTime` are `buy` and `sell` lockup. 

Yes description is need to be added in the code specifically at the declaration of `TimeRestriction`. while you can read this into the `modifywhitelist()` function params

Reply:
I added the following comment next to the struct definition for a bit more clarity:
```
    // _fromTime is the time from which the _investor can send tokens
    // _toTime is the time from which the _investor can receive tokens
    // if allowAllWhitelistIssuances is TRUE, then _toTime is ignored when receiving tokens from the issuance address
    // if allowAllWhitelistTransfers is TRUE, then _toTime and _fromTime is ignored when sending or receiving tokens
    // in any case, any investor sending or receiving tokens, must have a _expiryTime in the future
```